credentials/jwt/jwt_token_file.go

```diff
@@ -0,0 +1,285 @@
+	return true
+}
+
+// getToken returns a valid JWT token, reading from file if necessary.
+// Implements pre-emptive refresh and caches errors with backoff.
+func (c *jwtTokenFileCallCreds) getToken(ctx context.Context) (string, error) {
+	c.mu.RLock()
+
+	if c.isTokenValid() {
+		token := c.cachedToken
+		shouldRefresh := c.needsPreemptiveRefresh()
+		c.mu.RUnlock()
+
+		if shouldRefresh {
+			c.triggerPreemptiveRefresh()
+		}
+		return token, nil
+	}
+
+	// if still within backoff period, return cached error to avoid repeated file reads
+	if c.cachedError != nil && time.Now().Before(c.nextRetryTime) {
+		err := c.cachedError
+		c.mu.RUnlock()
+		return "", err
+	}
+
+	c.mu.RUnlock()
+	// Token is expired or missing or the retry backoff period has expired. So
+	// refresh synchronously.
+	// NOTE: refreshTokenSync itself acquires the write lock
+	return c.refreshTokenSync(ctx, false)
+}
+
+// isTokenValid checks if the cached token is still valid.
+// Caller must hold c.mu.RLock().
+func (c *jwtTokenFileCallCreds) isTokenValid() bool {
+	if c.cachedToken == "" {
+		return false
+	}
+	return c.cachedExpiration.After(time.Now())
+}
+
+// needsPreemptiveRefresh checks if a pre-emptive refresh should be triggered.
+// Returns true if the cached token is valid but expires within 1 minute.
+// We only trigger pre-emptive refresh for valid tokens - if the token is invalid
+// or expired, the next RPC will handle synchronous refresh instead.
+// Caller must hold c.mu.RLock().
+func (c *jwtTokenFileCallCreds) needsPreemptiveRefresh() bool {
+	return c.isTokenValid() && time.Until(c.cachedExpiration) < time.Minute
+}
+
+// triggerPreemptiveRefresh starts a background refresh if needed.
+// Multiple concurrent calls are safe - only one refresh will run at a time.
+// The refresh runs in a separate goroutine and does not block the caller.
+func (c *jwtTokenFileCallCreds) triggerPreemptiveRefresh() {
+	go func() {
+		c.refreshMu.Lock()
+		defer c.refreshMu.Unlock()
+
+		// Re-check if refresh is still needed under mutex
+		c.mu.RLock()
+		stillNeeded := c.needsPreemptiveRefresh()
+		c.mu.RUnlock()
+
+		if !stillNeeded {
+			return // Another goroutine already refreshed or token expired
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		// Force refresh to read new token even if current one is still valid
+		_, _ = c.refreshTokenSync(ctx, true)
+	}()
+}
+
+// refreshTokenSync reads a new token from the file and updates the cache.  If
+// preemptiveRefresh is true, bypasses the validity check of the currently cached
+// token and always reads from file.
+// This is used for pre-emptive refresh to ensure new tokens are loaded even when
+// the cached token is still valid. If preemptiveRefresh is false, skips file read
+// when cached token is still valid, optimizing concurrent synchronous refresh calls
+// where one RPC may have already updated the cache while another was waiting on the lock.
```

Comment:
Nit: Here and elsewhere, please ensure that comment lines are wrapped at 80 cols.